sql/database.go

```diff
@@ -292,16 +330,194 @@ func Open(uri string, opts ...Opt) (*sqliteDatabase, error) {
+				zap.Int("target version", after),
+			)
+		} else {
+			logger.Info("applying migrations to temporary DB",
+				zap.Int("current version", before),
+				zap.Int("target version", after),
+			)
+		}
+		return db, config.schema.Migrate(logger, db, before, config.vacuumState)
+	}
+}
+
 func Version(uri string) (int, error) {
 	pool, err := sqlitex.Open(uri, sqlite.SQLITE_OPEN_READONLY, 1)
 	if err != nil {
 		return 0, fmt.Errorf("open db %s: %w", uri, err)
 	}
 	db := &sqliteDatabase{pool: pool}
-	defer db.Close()
-	return version(db)
+	v, err := version(db)
+	return v, errors.Join(err, db.Close())
+}
+
+// deleteDB deletes the database at the specified path by removing /path/to/DB* files.
+// If the database doesn't exist, no error is returned.
+// In addition to what DROP DATABASE does, this also removes the migration marker file.
+func deleteDB(path string) error {
+	// https://www.sqlite.org/tempfiles.html plus marker *_done
+	for _, suffix := range []string{"", "-journal", "-wal", "-shm", "_done"} {
+		file := path + suffix
+		if err := os.Remove(file); err != nil {
+			if errors.Is(err, os.ErrNotExist) {
+				continue
+			}
+			return fmt.Errorf("remove %s: %w", file, err)
+		}
+	}
+	return nil
+}
+
+// moveMigratedDB runs "VACUUM INTO" on the database at fromPath and
+// replaces the database at toPath with the vacuumed one. The database
+// at fromPath is deleted after the operation.
+func moveMigratedDB(config *conf, fromPath, toPath string) (err error) {
+	config.logger.Warn("finalizing migration by moving the migrated DB to the original path",
+		zap.String("fromPath", fromPath),
+		zap.String("toPath", toPath))
+	// Try to open the migrated DB before deleting the original one.
+	// If the migrated DB is being copied to the original path by another
+	// process, this will fail and the original database will not be deleted.
+	// We don't use the proper database schema here because the migrated DB
+	// may have been created with a different set of migrations.
+	db, err := Open("file:"+fromPath,
+		WithLogger(config.logger),
+		WithConnections(1),
+		WithTemp(),
+		WithIgnoreSchemaDrift())
+	if err != nil {
+		return fmt.Errorf("open migrated DB %s: %w", fromPath, err)
+	}
+	if err := deleteDB(toPath); err != nil {
+		return err
+	}
+	if err := db.vacuumInto(toPath); err != nil {
+		return errors.Join(err, db.Close())
+	}
+	// Open the vacuumed DB to avoid race condition when another process
+	// also tries to vacuum the migrated DB into the original path after
+	// we close the migrated DB.
+	origDB, err := Open("file:"+toPath,
+		WithLogger(config.logger),
+		WithConnections(1),
+		WithMigrationsDisabled(),
+		WithIgnoreSchemaDrift(),
+		withDisableIncompleteMigrationHandling())
+	if err != nil {
+		return fmt.Errorf("open vacuumed DB %s: %w", toPath, err)
+	}
+	defer func() {
+		err = errors.Join(err, origDB.Close())
+	}()
+	if err := db.Close(); err != nil {
+		return fmt.Errorf("close migrated DB %s: %w", fromPath, err)
+	}
```

Comment:
while here I see the value of joining two errors (because they are not necessarily causally related) if they are different I don't know which error belongs to which DB if the `defer`red function doesn't add that information.

Reply:
Fixed